File: TiaCodegen/Extensions/OperationOrSignalExtensions.cs
using TiaCodegen.Interfaces;

namespace DotNetProjects.TiaCodegen.Extensions
{
    public static class OperationOrSignalExtensions
    {
        public static T TryGetParent<T>(this IOperationOrSignal op) where T : IOperationOrSignal
        {
            var chk = op.Parent;
            while (chk != null)
            {
                if (chk is T)
                    return (T)chk;
                chk = chk.Parent;
            }
            return default(T);
        }
    }
}

File: TiaCodegen-ts/src/Extensions/OperationOrSignalExtensions.ts
import { IOperationOrSignal } from '../Interfaces/IOperationOrSignal';

export function tryGetParent<T extends IOperationOrSignal>(
    op: IOperationOrSignal,
    ctor: new (...args: any[]) => T,
): T | null {
    let chk = op.parent;
    while (chk !== null && chk !== undefined) {
        if (chk instanceof ctor) return chk as T;
        chk = chk.parent;
    }
    return null;
}
